### PR TITLE
Reflection_Engine: Final tidy up

### LINKED ISSUE #2734: Reflection_Engine: Further review
Following merging of #2729 not all methods in Reflection_Engine need to continue residing there. That PR broke the core dependency between the BHoM_Engine and Reflection_Engine.

Further review is now needed to look at moving other methods, such as `IsPrototype`, etc. 

FYI: @adecler @IsakNaslundBh @al-fisher @alelom @pawelbaran

### PR DESCRIPTION
**Pull request by FraserGreenroyd**
### Issues addressed by this PR
Fixes #2734 

### Additional comments
Everything that remains/wasn't moved I consider to be a part of 'reflection' - the act of the code asking questions about itself in some fashion or another.

I added documentation for any file I touched that didn't have any.

I also added some unit tests because, why not? 😄 

For the `URL` unit test, I obtained all of the types from the BHoM repo, then checked all of their URLs to see which ones returned a `404` - indicating the URL generated is not valid for that type. I then removed those and produced the unit test on the working ones. The reason for this is that we then have a store of the working outputs that we don't want to change - and any subsequent PRs which fix or resolve the non-working URL generations will want to check the working ones haven't broken too. Hopefully that makes sense 😄

### CODE REVIEW

**Comment by FraserGreenroyd** [comment]
@BHoMBot check installer

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance

**Comment by FraserGreenroyd** [comment]
@BHoMBot check versioning

**Comment by FraserGreenroyd** [comment]
@BHoMBot check core
@BHoMBot check null-handling
@BHoMBot check serialisation
@BHoMBot check unit-tests

**Comment by FraserGreenroyd** [comment]
@BHoMBot check unit-tests

**Comment by FraserGreenroyd** [comment]
@BHoMBot check required

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance

File: Data_Engine/Query/Count.cs [hunk 1]
```diff
@@ -20,32 +20,57 @@
...
     {
+        /***************************************************/
+        /**** Interface Methods                         ****/
+        /***************************************************/
+
+        [PreviousVersion("6.0", "BH.Engine.Reflection.Query.ICount(System.Collections.Generic.List<System.Object>)")]
+        [Description("Get the count of the number of items in a list of objects.")]
+        [Input("list", "The list of items to obtain a count of.")]
+        [Output("count", "The number of items in the list.")]
+        public static int ICount<T>(this List<T> list)
```
Comment (IsakNaslundBh) [inline Data_Engine/Query/Count.cs:42]: Can't we just remove this method, and version it to the other one added on line 98.

Appreciate you are just moving things across, and already done some sensible clean-up, but think this additional step also could be done?

This is unless I am missing some obvious reason as to why we need to interface method, such as some difference in behaviour in something like Excel or similar, but have a hard time seeing that?

File: Data_Engine/Query/Count.cs [hunk 2]
```diff
@@ -56,12 +81,35 @@ public static int Count<T>(this Tree<T> tree)
...
+        public static int Count<T>(this List<T> list)
+        {
+            return list.Count;
+        }
+
+        /***************************************************/
+        /**** Private Methods                            ****/
+        /***************************************************/
+
+        private static int Count<T>(this object obj)
```
Comment (IsakNaslundBh) [inline Data_Engine/Query/Count.cs:107]: If we remove the interface method, we can get rid of this as well I think?

**Review by IsakNaslundBh — COMMENTED** [review]
Changes look good.

See comments on potential further simplification and removal of some methods that I cannot really see any point in keeping :)

**Comment by FraserGreenroyd** [comment]
@BHoMBot check compliance
@BHoMBot check unit-tests
@BHoMBot check versioning

**Comment by FraserGreenroyd** [comment]
@BHoMBot check null-handling
@BHoMBot check serialisation
@BHoMBot check core

**Review by IsakNaslundBh — APPROVED** [review]
Codechanges looks good and make sense.

Made a quick test of calling the Count method from exce, including checking the versioning functioning, and no issues to report.

Approved

**Comment by FraserGreenroyd** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**